File: webclient/src/io/Jagfile.ts
import BZip2 from '#/io/BZip2.js';
import Packet from '#/io/Packet.js';

export default class Jagfile {
    static genHash = (name: string): number => {
        let hash: number = 0;
        name = name.toUpperCase();
        for (let i: number = 0; i < name.length; i++) {
            hash = (hash * 61 + name.charCodeAt(i) - 32) | 0; // wtf?
        }
        return hash;
    };

    // constructor
    buffer: Uint8Array;
    compressedWhole: boolean;
    fileCount: number;
    fileHash: number[];
    fileUnpackedSize: number[];
    filePackedSize: number[];
    fileOffset: number[];
    fileUnpacked: Uint8Array[] = [];

    constructor(src: Uint8Array) {
        let data: Packet = new Packet(new Uint8Array(src));
        const unpackedSize: number = data.g3;
        const packedSize: number = data.g3;

        if (unpackedSize === packedSize) {
            this.buffer = src;
            this.compressedWhole = false;
        } else {
            this.buffer = BZip2.decompress(src.subarray(6), unpackedSize, true);
            data = new Packet(new Uint8Array(this.buffer));
            this.compressedWhole = true;
        }

        this.fileCount = data.g2;
        this.fileHash = [];
        this.fileUnpackedSize = [];
        this.filePackedSize = [];
        this.fileOffset = [];

        let offset: number = data.pos + this.fileCount * 10;
        for (let i: number = 0; i < this.fileCount; i++) {
            this.fileHash.push(data.g4);
            this.fileUnpackedSize.push(data.g3);
            this.filePackedSize.push(data.g3);
            this.fileOffset.push(offset);
            offset += this.filePackedSize[i];
        }
    }

    read(name: string): Uint8Array | null {
        const hash: number = Jagfile.genHash(name);
        const index: number = this.fileHash.indexOf(hash);
        if (index === -1) {
            return null;
        }
        return this.readIndex(index);
    }

    readIndex(index: number): Uint8Array | null {
        if (index < 0 || index >= this.fileCount) {
            return null;
        }

        if (this.fileUnpacked[index]) {
            return this.fileUnpacked[index];
        }

        const offset: number = this.fileOffset[index];
        const length: number = offset + this.filePackedSize[index];
        const src: Uint8Array = new Uint8Array(this.buffer.subarray(offset, offset + length));
        if (this.compressedWhole) {
            this.fileUnpacked[index] = src;
            return src;
        } else {
            const data: Uint8Array = BZip2.decompress(src, this.fileUnpackedSize[index], true);
            this.fileUnpacked[index] = data;
            return data;
        }
    }
}

File: javaclient/src/main/java/jagex2/io/Jagfile.java
package jagex2.io;

import deob.ObfuscatedName;

@ObfuscatedName("yb")
public class Jagfile {

	@ObfuscatedName("yb.e")
	public byte[] data;

	@ObfuscatedName("yb.f")
	public int fileCount;

	@ObfuscatedName("yb.g")
	public int[] fileHash;

	@ObfuscatedName("yb.h")
	public int[] fileUnpackedSize;

	@ObfuscatedName("yb.i")
	public int[] filePackedSize;

	@ObfuscatedName("yb.j")
	public int[] fileOffset;

	@ObfuscatedName("yb.k")
	public boolean unpacked;

	public Jagfile(byte[] src) {
		this.unpack(src);
	}

	@ObfuscatedName("yb.a([BB)V")
	public void unpack(byte[] src) {
		Packet buf = new Packet(src);
		int packedSize = buf.g3();
		int unpackedSize = buf.g3();

		if (unpackedSize == packedSize) {
			this.data = src;
			this.unpacked = false;
		} else {
			byte[] temp = new byte[packedSize];
			BZip2.decompress(temp, packedSize, src, unpackedSize, 6);
			this.data = temp;

			buf = new Packet(this.data);
			this.unpacked = true;
		}

		this.fileCount = buf.g2();
		this.fileHash = new int[this.fileCount];
		this.fileUnpackedSize = new int[this.fileCount];
		this.filePackedSize = new int[this.fileCount];
		this.fileOffset = new int[this.fileCount];

		int pos = buf.pos + this.fileCount * 10;
		for (int i = 0; i < this.fileCount; i++) {
			this.fileHash[i] = buf.g4();
			this.fileUnpackedSize[i] = buf.g3();
			this.filePackedSize[i] = buf.g3();
			this.fileOffset[i] = pos;
			pos += this.filePackedSize[i];
		}
	}

	@ObfuscatedName("yb.a(Ljava/lang/String;[B)[B")
	public byte[] read(String name, byte[] dst) {
		int hash = 0;
		String upper = name.toUpperCase();
		for (int i = 0; i < upper.length(); i++) {
			hash = hash * 61 + upper.charAt(i) - 32;
		}

		for (int i = 0; i < this.fileCount; i++) {
			if (this.fileHash[i] == hash) {
				if (dst == null) {
					dst = new byte[this.fileUnpackedSize[i]];
				}

				if (this.unpacked) {
					for (int j = 0; j < this.fileUnpackedSize[i]; j++) {
						dst[j] = this.data[this.fileOffset[i] + j];
					}
				} else {
					BZip2.decompress(dst, this.fileUnpackedSize[i], this.data, this.filePackedSize[i], this.fileOffset[i]);
				}

				return dst;
			}
		}

		return null;
	}
}
